HPC-9232 & HPC-9238: Upgrade to Node.js v20 and NPM v10

Pl217 opened the pull request:
The opportunity to upgrade dev dependencies is taken.

Pl217 committed:
HPC-9238: Upgrade to Node.js v20 in Docker image

Pl217 committed:
HPC-9238: Specify minimal version of Node.js v20

We still keep v18 as a valid option, since Dependabot
runner image still uses it
https://github.com/dependabot/dependabot-core/blob/main/npm_and_yarn/Dockerfile

Pl217 committed:
HPC-9232: Specify minimal version of NPM v10

We still keep v9 as a valid option, since
GitHub runner images haven't moved to v10 yet

Pl217 committed:
HPC-9238: Use Node.js v20 in GitHub CI

Pl217 committed:
Update GitHub actions to versions that use `node20` as default runtime

CodeQL Actions are yet to release v3, which will run on `node20`
https://github.com/github/codeql-action/issues/1959

Pl217 committed:
⬆️ Upgrade `@types/node` to `v20.8.10`

Pl217 committed:
Update ESLint to `v8.52.0`

Pl217 committed:
⬆️ Upgrade `lint-staged` to `v15.0.2`

Pl217 committed:
⬆️ Upgrade Prettier to `v3.0.3`

In v3, Prettier changed the default value for `trailingComma` to `all`.
To keep the old behavior, we need `trailingComma` set to `es5`.
https://prettier.io/blog/2023/07/05/3.0.0.html#breaking-changes

Pl217 committed:
Format file after upgrading to Prettier v3

This commit is output of running command `npx prettier -w .`

Pl217 committed:
Update to Typescript 5.2

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 committed:
OPS-9737: Upgrade minimal Node.js v18 version

Node.js v18.18.2 has multiple security fixes,
so we don't want earlier versions

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
